### PR TITLE
NestedFolders: Improve performance of Browse Dashboards by loading one page at a time

### LINKED ISSUE #68450: NestedFolders: Paginate Browse view
To help improve the loading speed of Browse Dashboards on instances with many dashboards and folders, we should load smaller pages of data initially and load more as the user scrolls down.

 - Should initially load `n` items for the root Browse view
 - As user scrolls, load additional `n`
 - If a folder with more than `n` items is expanded, it should also be paginated/lazy-loaded in the same way

### PR DESCRIPTION
**Pull request by joshhunt**
**What is this feature?**

 - 'Lazy loading' for dashboard browse items at the root view level. This works at both the `/dashboards` view, and for a folder view.
 - Loads 50-ish items initially (a page of dashboards is loaded if a all folders are loaded)
 - Loads additional pages as user scrolls down the page
 - Nested folders will load 999-ish children and are not yet paginated - the root view is where our main performance problems are

**Why do we need this feature?**

 - To fix performance issues with loading the many folders are the root view

**Special notes for your reviewer:**

 - Paginating this data is quite tricky because we don't have a unified API for 'items in folders'. Listing folders and dashboards come from two different APIs
   - So, we must maintain state about what previous pages have been loaded so we can figure out where to make the next request.
 - We use `react-window-infinite-loader` to handle the lazy loading.
   -  When generating the 'flatTree' list, if the children has not fully loaded we add 50 dummy 'pagination-placeholders' to the list. When those items are nearly scrolled into view, a request is triggered to load the next page of items.

Fixes https://github.com/grafana/grafana/issues/68450

### CODE REVIEW

**Review by ashharrison90 — COMMENTED** [review]
re: debounce behaviour... my smooth brain can't quite figure out how the old search table does this, but it manages to `await` the results without debouncing which might actually be nicer in this case? might be worth taking a look at the for inspiration

**Review by ashharrison90 — COMMENTED** [review]
looks like there's a bug when moving (probably deleting as well) an item that's not on the first page. steps to repro:

- have more than one page of results
- scroll down until the second page has loaded
- move an item inside another folder
- every item in the second page changes to loading and doesn't actually attempt to load until you scroll

couple of minor nits as well

File: public/app/features/browse-dashboards/state/hooks.ts [hunk 1]
```diff
@@ -134,9 +135,33 @@ function createFlatTree(
...
   const items = folderUID
-    ? (isOpen && childrenByUID[folderUID]) || [] // keep seperate lines
-    : rootItems;
+    ? isOpen && collection?.items // keep seperate lines
+    : collection?.items;
+
+  let children = (items || []).flatMap((item) => mapItem(item, folderUID, level));
+
+  if (level === 0 && collection && !collection.isFullyLoaded) {
+    console.log('collection is not fully loaded', collection);
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/features/browse-dashboards/state/hooks.ts:147]: gah! i looked so hard for all of these!

**Comment by joshhunt** [comment]
Great catch @ashharrison90. I did test scenarios like this, but you discovered an edge case. I think this only happens if you're scrolled just enough to see the second page of results. If a third page of results is visible, this won't happen (though you do get other weirdness that I think is just unavoidable at the moment...)

**Comment by joshhunt** [comment]
should be fixed @ashharrison90 if you could check it out again?

**Review by ashharrison90 — APPROVED** [review]
think we've still got a bit of simplifying/refactoring to do going forward to make this more easily understandable

**but** don't let me dampen the mood... this is complicated stuff, and this is a great foundation to build upon. the behaviour looks good to me, great job with this josh 👏👏👏👏

**Review by mckn — APPROVED** [review]
LGTM!

**Pull request merged**